trie/stacktrie.go

```diff
@@ -345,68 +360,44 @@ func (t *StackTrie) hash(st *stNode, path []byte) {
 		n := shortNode{Key: hexToCompactInPlace(st.key), Val: valueNode(st.val)}
 
 		n.encode(t.h.encbuf)
-		encodedNode = t.h.encodedBytes()
+		blob = t.h.encodedBytes()
 
 	default:
 		panic("invalid node type")
 	}
 
 	st.typ = hashedNode
 	st.key = st.key[:0]
-	if len(encodedNode) < 32 {
-		st.val = common.CopyBytes(encodedNode)
+
+	// Skip committing if the non-root node if the size is smaller than 32 bytes.
+	if len(blob) < 32 && len(prefix) > 0 {
+		st.val = common.CopyBytes(blob)
 		return
 	}
-
 	// Write the hash to the 'val'. We allocate a new val here to not mutate
-	// input values
-	st.val = t.h.hashData(encodedNode)
-	if t.writeFn != nil {
-		t.writeFn(path, common.BytesToHash(st.val), encodedNode)
+	// input values.
+	st.val = t.h.hashData(blob)
+
+	// Commit the trie node if the writer is configured.
+	if t.options.Writer != nil {
+		t.options.Writer(prefix, common.BytesToHash(st.val), blob)
 	}
 }
 
-// Hash returns the hash of the current node.
-func (t *StackTrie) Hash() (h common.Hash) {
-	st := t.root
-	t.hash(st, nil)
-	if len(st.val) == 32 {
-		copy(h[:], st.val)
-		return h
-	}
-	// If the node's RLP isn't 32 bytes long, the node will not
-	// be hashed, and instead contain the  rlp-encoding of the
-	// node. For the top level node, we need to force the hashing.
-	t.h.sha.Reset()
-	t.h.sha.Write(st.val)
-	t.h.sha.Read(h[:])
-	return h
+// Hash will firstly hash the entire trie if it's still not hashed
+// and then commit all nodes to the associated database. Actually most
+// of the trie nodes MAY have been committed already. The main purpose
+// here is to commit the root node.
+func (t *StackTrie) Hash() common.Hash {
+	n := t.root
+	t.hash(n, nil)
+	return common.BytesToHash(n.val)
```

Comment:
This change does not only modify the constructor, you are actually (as far as I can tell) modifying how the stacktrie will behave with tiny tries, e..g a trie with only one value, embedded into the root fullnode. The previous code will force-hash it, the current will not (?)

Reply:
Could also add to the docs for both Hash and Commit
``
// For stacktrie, Hash and Commit are functionally identical
```